recipes/default.rb

```diff
@@ -15,14 +15,14 @@
   node['rackspace_users']['data_bag'],
   node['rackspace_users']['data_bag_item']).to_hash.select { |user, user_data| user != 'id' }
 
-node_tags = node['rackspace_users']['node_tags']
+node_groups = node['rackspace_users']['node_groups']
 
 groups = {}
 
 users.each do |username, user_data|
-  ## The action (if defined in the data bag) should be overwritten to 'remove' if the user's node tags and node's tags don't intersect.
-  ## If the user doesn't declare any tags then it is assumed that is present on all nodes.
-  user_data['action'] = 'remove' if user_data['node_tags'] && (user_data['node_tags'] & node_tags).empty?
+  ## The action (if defined in the data bag) should be overwritten to 'remove' if not all the user's node_groups are in node's node_groups.
+  ## If the user doesn't declare node_groups then it is assumed that is present on all nodes.
+  user_data['action'] = 'remove' if user_data['node_groups'] && !(user_data['node_groups'] - node_groups).empty?
```

Comment:
Good point @seanrettig . I don't believe that we can have something that will work for every case without making the logic (or the data structure and the logic) overly complicated which is not what I want to do. As simple as possible is better.

What I was thinking is that we can just keep the 'any' logic and then we can achieve most of what we want by leveraging the way we define the node_groups. Here are the scenarios I think will be the most common cases and we can implement them just with the 'any' logic. What we can additionally do is add these examples to the README to give guidance as to how this cookbook can be consumed effectively:
 
##### user must be on all servers regardless of role or environment:
```
node_groups should not be defined for the user
```
##### user must be on all servers on environment X:
On any node:
`node[‘rackspace_users’][‘node_groups’] = [ node.chef_environment ]`
On the `users` data bag item:
```
“user”: {
  “node_groups”: [ “X” ]
}
```
##### user must be on all servers running role X on any environment:
On the node with role ‘X’:
`node[‘rackspace_users’][‘node_groups’] = [ ‘X’ ]`
On the ‘users’ data bag item:
```
“user”: {
  “node_groups”: [ “X” ]
}
```
##### user must be on all servers running role X or role Y on any environment:
On the node with role ‘X’:
`node[‘rackspace_users’][‘node_groups’] = [ ‘X’ ]`
On the node with role ‘Y’:
`node[‘rackspace_users’][‘node_groups’] = [ ‘Y’ ]`
On the ‘users’ data bag item:
```
“user”: {
  “node_groups”: [ “X” , “Y” ]
}
```
##### user must be on all servers running role X only on environment Y (here is the 'tricky' part):
On the node with role ‘X’:
`node[‘rackspace_users’] = [ “X_#{node.chef_environment}” ]`
On the ‘users’ data bag item:
```
“user”: {
  “node_groups”: [ “X_Y” ]
}
```
##### user must be on all servers running role X or role Y only on environment Z:
On the node with role ‘X’:
`node[‘rackspace_users’][‘node_groups’] = [ “X_#{node.chef_environment}” ]`
On the node with role ‘Y’:
`node[‘rackspace_users’][‘node_groups’] = [ “Y_#{node.chef_environment}” ]`
On the ‘users’ data bag item:
```
“user”: {
  “node_groups”: [ “X_Z” , “Y_Z” ]
}
```
##### user must be on all servers running role X only on environment Y and servers running role Z but only on environment K:
On the node with role ‘X’:
`node[‘rackspace_users’]['node_groups'] = [ “X_#{node.chef_environment}” ]`
On the node with role ‘Z’:
`node[‘rackspace_users’]['node_groups'] = [ “Z_#{node.chef_environment}” ]`
On the ‘users’ data bag item:
```
“user”: {
  “node_groups”: [ “X_Y” , “Z_K” ]
}
```

will also like @martinb3 input on this one. If you guys think that this is reasonable and covers most cases then I can go ahead and implement it.